File: app/src/main/java/home/oleg/placesnearme/di/modules/NetworkModule.java
package home.oleg.placesnearme.di.modules;

import dagger.Module;
import dagger.Provides;
import home.oleg.placesnearme.BuildConfig;
import home.oleg.placesnearme.network.config.NetworkConfig;
import io.reactivex.annotations.NonNull;

@Module
public final class NetworkModule {

    @Provides
    @NonNull
    public NetworkConfig provideNetwowkConfig() {
        return new NetworkModule.MyNetworkConfig();
    }

    private static class MyNetworkConfig implements NetworkConfig {
        @Override
        public String baseUrl() {
            return BuildConfig.BASE_URL;
        }

        @Override
        public String clientSecret() {
            return BuildConfig.CLIENT_SECRET;
        }

        @Override
        public String clientId() {
            return BuildConfig.CLIENT_ID;
        }

        @Override
        public String apiVersion() {
            return BuildConfig.API_VERSION;
        }
    }
}

File: app/src/main/java/home/oleg/placesnearme/app/di/modules/NetworkModule.kt
package home.oleg.placesnearme.app.di.modules

import androidx.annotation.NonNull
import dagger.Module
import dagger.Provides
import home.oleg.placesnearme.BuildConfig
import home.oleg.placesnearme.network.config.NetworkConfig
import home.oleg.placesnearme.network.service.FourSquareAPIFactory
import home.oleg.placesnearme.network.service.IFourSquareAPI

@Module
object NetworkModule {

    @JvmStatic
    @Provides
    internal fun provideApi(networkConfig: NetworkConfig): IFourSquareAPI {
        return FourSquareAPIFactory.create(networkConfig)
    }

    @JvmStatic
    @Provides
    internal fun provideNetworkConfig(): NetworkConfig {
        return NetworkModule.MyNetworkConfig()
    }

    private class MyNetworkConfig : NetworkConfig {
        override fun baseUrl(): String {
            return BuildConfig.BASE_URL
        }

        override fun clientSecret(): String {
            return BuildConfig.CLIENT_SECRET
        }

        override fun clientId(): String {
            return BuildConfig.CLIENT_ID
        }

        override fun apiVersion(): String {
            return BuildConfig.API_VERSION
        }
    }
}
